src/protocolsupport/protocol/packet/middleimpl/serverbound/play/v_pe/CustomPayload.java

```diff
@@ -16,7 +17,7 @@ public CustomPayload(ConnectionImpl connection) {
 	@Override
 	public void readFromClientData(ByteBuf clientdata) {
 		tag = StringSerializer.readString(clientdata, connection.getVersion(), 20);
-		data = MiscSerializer.readAllBytesSlice(clientdata);
+		data = Unpooled.wrappedBuffer(ArraySerializer.readVarIntByteArray(clientdata));
```

Comment:
There exists a slice method

Reply:
there is, but im all weird about ByteBuf reference tracking now. at this point (given i cant reason about reference tracking from here), it just seems easier to do a heap copy at this point. slices are a direct reference, so this is a rare point where heap ByteBuf is probably ideal.

Reply:
Not really, since the data is read and transformed just after that.